internal/command/cliconfig/oci_registry.go

```diff
@@ -0,0 +1,373 @@
+				fmt.Sprintf("The oci_registries block at %s must not have any labels.", block.Pos()),
+			))
+		}
+
+		cfg := &OCIRegistries{}
+
+		// The structure _inside_ an oci_registry block is straightforward
+		// enough for us to decode it using the HCL 1 decoder without permitting
+		// structures that would be super hard to migrate to HCL 2 later.
+		type blockInnards struct {
+			CredentialHelper        string            `hcl:"credential_helper"`
+			DomainCredentialHelpers map[string]string `hcl:"domain_credential_helpers"`
+			// The "StaticCredentials" field is intentionally not supported here
+			// because we don't want to encourage folks to duplicate their
+			// static credentials from their Docker CLI configuration. That field
+			// is there only to allow us to automatically discover credentials that
+			// were previously written by "docker login", for folks who have both
+			// OpenTofu and Docker CLI installed.
+		}
+		var innards blockInnards
+		if err := hcl.DecodeObject(&innards, block.Val); err != nil {
+			// HCL 1 doesn't tend to generate very high-quality error messages, but
+			// we'll accept it as good enough with some extra context added on here.
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Invalid oci_registries block",
+				fmt.Sprintf("The oci_registries block at %s has invalid content: %s.", block.Pos(), err),
+			))
+		}
+
+		ret = ret.merge(cfg)
+	}
+
+	return ret, diags
+}
+
+// impliedOCIRegistrySettingsFromDockerConfig makes a best effort to construct
+// an [OCIRegistries] based on the configuration files and environment variables
+// normally used and updated by Docker CLI.
+//
+// This directly accesses the relevant environment variables and files on disk
+// to encapsulate all of the Docker-specific details. Use
+// [impliedOCIRegistrySettingsFromDockerConfigFile] instead if you've already
+// got the source code of a Docker CLI config file loaded into a byte slice.
+//
+// Returns nil if there does not appear to be any ambient Docker CLI configuration
+// available.
+//
+// The goal here is to give a better "out of box" experience for anyone who
+// already has Docker CLI installed and configured to interact with the
+// OCI registries they want to use, and in particular so that we can make
+// use of credentials previously issued by "docker login" to avoid our users
+// needing to duplicate those secrets in another place.
+//
+// This is not a fully-comprehensive implementation of the Docker CLI language,
+// and since the evolution of the Docker CLI language is not under OpenTofu's
+// control this function prefers to ignore anything unexpected rather than
+// returning an error. It does, however, produce some log output for certain
+// problems to make it easier to handle bug reports that suggest our
+// interpretation of this language is not sufficiently compatible with the
+// official Docker/Moby implementation.
+func impliedOCIRegistrySettingsFromDockerConfig() *OCIRegistries {
```

Comment:
We can easily delete this fallback of trying to infer the settings from Docker CLI if you think it's not helpful.

I don't think we ought to get into all the complexity of supporting the configuration formats of many different pieces of software, or of implementing a special meta-plugin model that would really just be for discovering which _other_ plugins (the Docker credentials helpers) to use.

I had assumed that Docker CLI was a common enough starting point to be worth a small amount of code for special support, but if you don't believe that to be true then I think we should just delete this and make "explicit mode" the only option.

Reply:
Interestingly, the Podman documentation suggests that Podman itself also falls back to the Docker CLI configuration when it can't find its own configuration files, and so what I did here is at least consistent with what Podman did. :man_shrugging: 

> Path of the authentication file. Default is `${XDG_RUNTIME_DIR}/containers/auth.json` on Linux, and `$HOME/.config/containers/auth.json` on Windows/macOS. The file is created by [podman login](https://docs.podman.io/en/latest/markdown/podman-login.1.html#). If the authorization state is not found there, `$HOME/.docker/config.json` is checked, which is set using `docker login`.

The last part also implies that the motivation was essentially the same as mine: to make any static credentials that were already issued by `docker login` visible to Podman, so that the user doesn't need to duplicate them in two places when they are using both pieces of software together.

(My uninformed guess would be that lots of folks still use `docker build`/`docker buildx`/`docker push` for building and publishing even if they don't use the Docker server to run their containers, and so other software in this ecosystem feels justified in supporting just enough of that tool's configuration language to benefit from the already-established registry credentials.)

Reply:
Having poked around some more, it seems that:

1. This `containers/auth.json` file has a compatible format to Docker's `config.json`.
2. This file path has also been implemented by other software, e.g. https://github.com/google/go-containerregistry/issues/1145

Therefore it seems like it would be a plausible compromise to check both locations and try to read from both of them in the same way. The actual decoding function would be shared between them, so it's not much work to support one more search location for that file.

Just for the sake of having something concrete to discuss, then, I'm going to change this so that it'll first search for the `containers/auth.json` file, and then treat the `.docker/config.json` one as a Docker-CLI-specific fallback for that, treating both of them as having the same file format.

Since other software seems to have rallied around the path that Podman chose, and that path seems to be intentionally generic (it doesn't have the word "podman" in it anywhere), I hope that is a sufficient compromise to avoid all the complexity of making this fully general/modular.